pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+
+        Parameters
+        ----------
+        size : size_t
+            the number of items in the list
+
+        type : t_type
+            The type of the new list's elements
+
+        elemnts : void*
+            A pointer on a table containing the list's elements
+
+        Returns
+        =======
+        list : t_list
+*/
+{
+    t_list *list;
+    size_t tsize = tSizes[type];
+
+    if (!(list = (t_list*)malloc(sizeof(t_list))))
+        return NULL;
+    list->size = size;
+    list->type = type;
+    list->capacity = DEFAULT_CAP;
+    while (list->capacity <= list->size)
+        list->capacity *= 2;
+    if (!(list->elements = malloc(list->capacity * tsize)))
+        return NULL;
+    if (elemnts)
+        memcpy(list->elements, elemnts, tsize * list->size);
+    return list;
+}
+
+void     free_list(t_list **list)
+/*
+        Deallocate a t_list object
+
+        Parameters
+        ----------
+        list : t_list**
+            a double pointer on the list to be freed
+*/
+{
+    free((*list)->elements);
+    free(*list);
+    *list = NULL;
+}
+
+void    extend(t_list* list1, t_list* list2)
+{
+    size_t totalSize = list1->size + list2->size;
+    size_t tsize = tSizes[list1->type];
+    char *elements = list1->elements;
+
+    if (totalSize >= list1->capacity)
+    {
+        while (list1->capacity <= totalSize)
+            list1->capacity *= 2;
+        elements = realloc(elements, list1->capacity * tsize);
+        list1->elements = elements;
+    }
+    memcpy(&elements[list1->size * tsize], list2->elements, list2->size * tsize);
+    list1->size = totalSize;
+}
+
+void     clear(t_list* list)
+/*
+        Clears a t_list without deallocation
+
+        Parameters
+        ----------
+        list : t_list**
+            a pointer on the list to be freed
+*/
+{
+    list->size = 0;
+}
+
+t_list*  copy(t_list* list)
+/*
+        Creates a copy t_list object from another
+
+        Parameters
+        ----------
+        list : t_list*
+            a pointer on the list to be copied
+        
+        Returns
+        =======
+        list : t_list
+*/
+{
+    return (allocate_list(list->size, list->type, list->elements));
+}
+
+size_t   count(t_list* list, void *item)
+/*
+        Counts the number of occurrences of a specific element
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list where to look for occurrences
+
+        item : void*
+            The element to look for
+
+        Returns
+        =======
+        count : size_t
+*/
+{
+    size_t index = 0;
+    int count = 0;
+    char *elements = list->elements;
+    while (index < list->size)
+    {
+        if (list->type == lst_list)
+        {
+            t_list *tmp = ((t_list**)elements)[index];
+            if (!(memcmp(item, tmp, sizeof(t_list))))
+                count += 1;
+        }
+        else if (!(memcmp(item, &(elements[index * tSizes[list->type]]), tSizes[list->type])))
+                count += 1;
```

Comment:
Does `list.count` use `==` comparison or `is` comparison? This looks like an `is` comparison. Do you think that this is a problem? Or only if we add more complex objects such as classes later?

Reply:
well for now we only support literals and lists themselves, as such this should be a == comparison as it compares he raw binary data, in the case of lists themselves I still want to add a functionality to iterate through the elements.
For more complex types tho I was thinking maybe we could a function table that holds type specific comparison functions,  with the indexes being the value of the enum corresponding to the specific type. 
something like this
```c
typedef (int)(compare_func)(t_list*, void*) compFuncs;
compFuncs[] = {compare_typeA, compare_typeB, ...}
```
I m not sure if this is the best way to do it.

Reply:
I am also not sure. I think it will depend how we plan to manage classes.

Reply:
I think this should be ok for now. As you say this should be sufficient for literals. Arrays don't have a comparison operator returning a scalar, and we haven't implemented magic methods for classes yet so it uses the default comparison (equivalent to `is`)

Reply:
I think that will be the case

Reply:
@azouiten This function is not working as it should be. There was is an issue when I tried to count a elements in multidimensional lists.
Consider this:
```python
b = [[1, 2], [3, 4], [5, 6]]
print(b.count([3, 4]))
```
The code above translates to this C code:
```C
    t_list *b = NULL;
    t_list *Dummy_0000 = NULL;

    b = allocate_list(3, lst_list, (t_list*[]){
        allocate_list(1, lst_int64, (int64_t[]){
            INT64_C(1), INT64_C(2)
        }), 
        allocate_list(1, lst_int64, (int64_t[]){
            INT64_C(3), INT64_C(4)
        }), 
        allocate_list(1, lst_int64, (int64_t[]){
            INT64_C(5), INT64_C(6)
        })
    });

    Dummy_0000 = allocate_list(1, lst_int64, (int64_t[]){
        INT64_C(1)
    });

    printf("%"PRId64"\n", count(b, Dummy_0000));
```
This code will always print `0`, because if you check line 127 it will only compare the bytes of the two structures pointed to by `b` and `Dummy_0000` and not the actual data in elements.